experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,497 @@
+    print(io, "Reductive group ", G.group[1], G.group[2])
+end
+
+function reductive_group(sym::Symbol, m::Int, F::Field)
+    return ReductiveGroup(sym,m,F)
+end
+
+function reductive_group(sym::Symbol, m::Int, R::MPolyRing)
+    return ReductiveGroup(sym,m,R)
+end
+
+group(G::ReductiveGroup) = G.group
+reynolds_operator(G::ReductiveGroup) = G.reynolds_operator
+group_ideal(G::ReductiveGroup) = G.group_ideal
+canonical_representation(G::ReductiveGroup) = G.canonical_representation
+natural_representation(G::ReductiveGroup) = G.canonical_representation
+
+#####################
+#Representation of Reductive Groups. Embeds SLm in GLn. 
+#####################
+
+mutable struct RepresentationReductiveGroup
+    group::ReductiveGroup
+    rep_mat::AbstractAlgebra.Generic.MatSpaceElem
+    sym_deg::Tuple{Bool, Int}
+    reynolds_v::Function
+    
+    #representation of group G over symmetric degree d
+    function RepresentationReductiveGroup(G::ReductiveGroup, d::Int)
+        R = new()
+        R.group = G
+        R.rep_mat = rep_mat_(G, d)
+        R.sym_deg = (true, d)
+        R.reynolds_v = reynolds_v_slm
+        return R
+    end
+    
+    #matrix M is the representation matrix. does not check M.
+    function RepresentationReductiveGroup(G::ReductiveGroup, M::AbstractAlgebra.Generic.MatSpaceElem)
+        @req base_ring(M) == base_ring(G.group_ideal) "Group ideal and representation matrix must have same parent ring"
+        R = new()
+        R.group = G
+        R.rep_mat = M
+        R.sym_deg = (false, 0)
+        R.reynolds_v = reynolds_v_slm
+        return R
+    end
+end
+
+representation_reductive_group(G::ReductiveGroup, d::Int) = RepresentationReductiveGroup(G, d)
+representation_reductive_group(G::ReductiveGroup, M::AbstractAlgebra.Generic.MatSpaceElem) = RepresentationReductiveGroup(G,M)
+
+function representation_on_forms(G::ReductiveGroup, d::Int)
+    return RepresentationReductiveGroup(G, d)
+end
+
+representation_matrix(R::RepresentationReductiveGroup) = R.rep_mat
+reductive_group(R::RepresentationReductiveGroup) = R.group
+
+#returns dimension n
+function vector_space_dimension(R::RepresentationReductiveGroup)
+    return ncols(R.rep_mat)
+end
+
+function Base.show(io::IO, R::RepresentationReductiveGroup)
+    println(io, "Representation of ", R.group.group[1], R.group.group[2])
+    if R.sym_deg[1]
+        print(io, "over symmetric forms of degree ", R.sym_deg[2])
+    else
+        println(io, "with representation matrix")
+        show(io, R.rep_mat)
+    end
+end
+
+###############
+
+#computes the representation matrices of SL_m acting over m-forms of symmetric degree sym_deg
+function rep_mat_(G::ReductiveGroup, sym_deg::Int)
+    G.group[1] == :SL || error("Only implemented for SLm")
+    m = G.group[2]
+    n = binomial(m + sym_deg - 1, m - 1)
+    mixed_ring, t, z = polynomial_ring(G.field, "t"=> 1:m, "z"=> (1:m, 1:m))
+    group_mat = matrix(mixed_ring, z)
+    new_vars = group_mat*t
```

Comment:
one step after the other, but since we discussed it and it shouldn't be forgotten: it may be better to rewrite this code to use nested polynomial rings. Say like this (untested, and even if I got it right, it'll require additional changes elsewhere in the code):
```suggestion
    R = base_ring(G.group_ideal) # TODO: probably should have a getter function for this
    mixed_ring, t = polynomial_ring(R, "t" => 1:m)
    group_mat = natural_representation(G)
    new_vars = group_mat*t
```